web_poet/testing/fixture.py

```diff
@@ -0,0 +1,169 @@
+import asyncio
+import datetime
+import json
+import logging
+import os
+import sys
+from pathlib import Path
+from typing import Any, Iterable, Optional, Type, TypeVar, Union
+
+import dateutil.parser
+import dateutil.tz
+import time_machine
+from itemadapter import ItemAdapter
+
+from web_poet import ItemPage
+from web_poet.serialization import (
+    SerializedDataFileStorage,
+    deserialize,
+    load_class,
+    serialize,
+)
+from web_poet.utils import ensure_awaitable
+
+logger = logging.getLogger(__name__)
+
+
+INPUT_DIR_NAME = "inputs"
+OUTPUT_FILE_NAME = "output.json"
+META_FILE_NAME = "meta.json"
+
+
+FixtureT = TypeVar("FixtureT", bound="Fixture")
+
+
+def _get_available_filename(template: str, directory: Union[str, os.PathLike]) -> str:
+    i = 1
+    while True:
+        result = Path(directory, template.format(i))
+        if not result.exists():
+            return result.name
+        i += 1
+
+
+class Fixture:
+    """Represents a directory containing one test."""
+
+    def __init__(self, path: Path) -> None:
+        self.path = path
+
+    @property
+    def type_name(self) -> str:
+        """The name of the type being tested."""
+        return self.path.parent.name
+
+    @property
+    def input_path(self) -> Path:
+        """The inputs subdirectory path."""
+        return self.path / INPUT_DIR_NAME
+
+    @property
+    def output_path(self) -> Path:
+        """The output file path."""
+        return self.path / OUTPUT_FILE_NAME
+
+    @property
+    def meta_path(self) -> Path:
+        """The metadata file path."""
+        return self.path / META_FILE_NAME
+
+    def is_valid(self) -> bool:
+        """Return True if the fixture file structure is correct, False otherwise."""
+        return self.input_path.is_dir() and self.output_path.is_file()
+
+    def get_page(self) -> ItemPage:
+        """Return the page object created from the saved input."""
+        cls = load_class(self.type_name)
+        if not issubclass(cls, ItemPage):
+            raise TypeError(f"{self.type_name} is not a descendant of ItemPage")
+        storage = SerializedDataFileStorage(self.input_path)
+        return deserialize(cls, storage.read())
+
+    def get_meta(self) -> dict:
+        """Return the test metadata."""
+        if not self.meta_path.exists():
+            return {}
+        return json.loads(self.meta_path.read_bytes())
+
+    def get_output(self) -> dict:
+        """Return the output from the recreated Page Object."""
+        po = self.get_page()
+        item = asyncio.run(ensure_awaitable(po.to_item()))
+        return ItemAdapter(item).asdict()
+
+    def get_expected_output(self) -> dict:
+        """Return the saved output."""
+        return json.loads(self.output_path.read_bytes())
+
+    @staticmethod
+    def _parse_frozen_time(meta_value: str) -> datetime.datetime:
+        """Parse and possibly fix the frozen_time metadata string."""
+        parsed_value = dateutil.parser.parse(meta_value)
```

Comment:
Is dateutil only used to allow flexible time formats? If so, can we drop this dependency, and make time formats more restricted?

Reply:
time-machine depends on it anyway.